frontend/frontend/honeylogger/__init__.py

```diff
@@ -0,0 +1,102 @@
+        """Logs a new login attempt associated with the current session.
+
+        :param username: The username used in the login attempt.
+        :param password: The password used in the login attempt.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def log_command(self, input: str) -> None:
+        """Logs a new command associated with the current session.
+
+        :param input: The raw string that was input to run the command.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def log_download(self,
+                     data: memoryview,
+                     file_type: str,
+                     source_address: IPAddress,
+                     source_url: Optional[str] = None,
+                     save_data: bool = True) -> None:
+        """Logs a new downloaded file associated with the current session.
+
+        :param data: The binary contents of the downloaded file.
+        :param file_type: The file type of the downloaded file. Should be a valid MIME type.
+            If unknown, `application/octet-stream` should be used.
+        :param source_address: The IP address of the downloaded file's source.
+        :param source_url: If known, the URL where the downloaded file originated from.
+        :param save_data: If true, the binary contents of the downloaded file will be saved,
+            if not, only metadata is saved.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def end(self) -> None:
+        """Marks the session as having ended."""
+        raise NotImplementedError
+
+
+class SSHSession(Session, Protocol):
+    """Representation of an attacker's SSH session while being connected to the honeypot."""
+
+    @abstractmethod
+    def log_pty_request(self, term: str,
+                        term_width_cols: int, term_height_rows: int,
+                        term_width_pixels: int, term_height_pixels: int) -> None:
+        """Logs an SSH pty request associated with the current session.
+
+        :param term: ``TERM`` environment variable value (e.g. vt100).
+        :param term_width_cols: Terminal width, characters (e.g. 80)
+        :param term_height_rows: Terminal height, rows (e.g. 24)
+        :param term_width_pixels: Terminal width, pixels (e.g. 640)
+        :param term_height_pixels: Terminal height, pixels (e.g. 480)
+        """
+        raise NotImplementedError
+
+
+def begin_ssh_session(src_address: IPAddress, src_port: int,
+                      dst_address: IPAddress, dst_port: int) -> SSHSession:
+    """Begins to log a new SSH session.
+
+    :param src_address: The IP address of the instigating origin of the session.
+    :param src_port: The port number at the instigating origin.
+    :param dst_address: The public IP address of the honeypot.
+    :param dst_port: The port at the honeypot.
+    :return: An established SSH session.
+    """
+
+    session = ConsoleLogSSHSession(src_address=src_address,
```

Comment:
Why do you need this proxy function? Wouldn't it work better to just call the constructor directly?

Reply:
We have this so we can easily change between a ConsleLogSSHSession and PostgresLogSSHSession object depending if we want to log to console or the database